src/Control/Concurrent/TOrderedQueue.hs

```diff
@@ -0,0 +1,106 @@
+    newTQueueIO,
+    retry,
+    tryReadTQueue,
+    writeTQueue,
+  )
+
+import Data.PQueue.Prio.Min (MinPQueue, insert, minViewWithKey)
+
+data SequenceId
+  = Unordered
+  | SequenceId !Natural
+  deriving stock (Eq, Ord, Show)
+
+-- | A value `a` annotated with a `SequenceId`
+data Sequenced a = Sequenced
+  { idx :: !SequenceId
+  , val :: !a
+  }
+  deriving stock (Show)
+
+instance Eq (Sequenced a) where
+  (==) = (==) `on` idx
+
+instance Ord (Sequenced a) where
+  (<=) = (<=) `on` idx
+
+-- | A queue in which messages written to the queue can be annotaed with a `SequenceId`, and messages read from
+-- the queue will always be seen in order by `SequenceId`, even if the messages are written out-of-order.
+-- Messages written to the queue annoted with `Unordered` will be read before any ordered messages and will be
+-- processed in FIFO order.
+data TOrderedQueue a
+  = TOrderedQueue
+      !(TVar Natural) -- Sequence counter indicating what the sequence id of the next message read should be
+      !(TVar (MinPQueue Natural a)) -- A min heap for holding messages annoted by `SequenceId`
+      !(TQueue a) -- A regular TQueue used for processing `Unordered` messages
+
+-- | Build and return a new `TOrderedQueue`
+newTOrderedQueue :: STM (TOrderedQueue a)
+newTOrderedQueue =
+  TOrderedQueue
+    <$> newTVar 0
+    <*> newTVar mempty
+    <*> newTQueue
+
+-- | `IO` version of `newTOrderedQueue`
+newTOrderedQueueIO :: IO (TOrderedQueue a)
+newTOrderedQueueIO = do
+  TOrderedQueue
+    <$> newTVarIO 0
+    <*> newTVarIO mempty
+    <*> newTQueueIO
+
+-- | Write an item, annotated by `Sequenced`, to an `TOrderedQueue`
+writeTOrderedQueue :: TOrderedQueue a -> Sequenced a -> STM ()
+writeTOrderedQueue (TOrderedQueue _ orderedQueueVar unorderedQueue) (Sequenced sid x) =
+  case sid of
+    Unordered -> writeTQueue unorderedQueue x
+    SequenceId i -> modifyTVar' orderedQueueVar (insert i x)
+
+-- | Read the next item from the `TOrderedQueue`.
+-- It will first return any `Unordered` items if one is available.
+-- If no `Unordered` items are available, it will return the next item from the
+-- ordered queue if the index matches that of the current sequence counter.
+-- Otherwise blocks until an `Unordered` item is written, or an item with the correct
+-- sequence id is written.
+readTOrderedQueue :: TOrderedQueue a -> STM a
+readTOrderedQueue otq@(TOrderedQueue seqVar orderedQueueVar unorderedQueue) =
+  tryReadTQueue unorderedQueue >>= \case
+    Just a -> pure a
+    Nothing -> do
+      orderedQueue <- readTVar orderedQueueVar
+      case minViewWithKey orderedQueue of
+        Nothing -> retry
+        Just ((i, a), rest) -> do
+          curSeq <- readTVar seqVar
+          case i `compare` curSeq of
+            -- If sequence number is less than current, it must be a duplicate,
+            -- so we discard it and read again
+            LT -> do
+              writeTVar orderedQueueVar rest
+              readTOrderedQueue otq
```

Comment:
No, this is all in an `STM` transaction, the state of the queues isn't able to change mid-transaction. It would always see the `Unordered` queue as empty until the transaction ends or a `retry` is hit. It will continue to take the min from the heap until it finds an element with a greater or equal index. Although, on a `retry`, all of those elements will be restored. Unfortunately this means they would be traversed again every time until the correct index is found.

However, there is probably a small optimization opportunity to recurse on only reading from the ordered queue to skip checking the ordered queue each time.

Reply:
Oh that's right, that makes sense to me. Thanks for explaining!